File: src/main/java/com/bjworld21/congress/service/RecipientGroups.java
package com.bjworld21.congress.service;

import java.util.LinkedHashSet;
import java.util.List;

public final class RecipientGroups {
    private RecipientGroups() {}
    public static final List<String> CODES = List.of("ALL_MEMBERS", "ALL_REGISTRANTS", "ALL_SUBMITTERS", "ALL_ACCEPTED");
    public static final String MEMBER_PREDICATE = """
            (#{group} = 'ALL_MEMBERS'
             OR (#{group} = 'ALL_REGISTRANTS' AND EXISTS (
                 SELECT 1 FROM pre_registrations p WHERE p.memberSeq = m.seq
                 AND p.conferenceSeq = m.conferenceSeq AND p.applicationStatus = 'SUBMITTED'))
             OR (#{group} = 'ALL_SUBMITTERS' AND EXISTS (
                 SELECT 1 FROM abstract_submissions s WHERE s.memberSeq = m.seq
                 AND s.conferenceSeq = m.conferenceSeq
                 AND s.status IN ('submitted', 'under_review', 'approved', 'rejected')))
             OR (#{group} = 'ALL_ACCEPTED' AND EXISTS (
                 SELECT 1 FROM abstract_submissions s WHERE s.memberSeq = m.seq
                 AND s.conferenceSeq = m.conferenceSeq AND s.status = 'approved')))
            """;
    public static List<String> validate(List<String> groups) {
        if (groups == null) return List.of();
        if (groups.stream().anyMatch(group -> group == null || !CODES.contains(group))) {
            throw new IllegalArgumentException("전체 수신 대상 구분이 올바르지 않습니다.");
        }
        return List.copyOf(new LinkedHashSet<>(groups));
    }
}

File: frontend/src/recipientGroups.ts
export const RECIPIENT_GROUPS = [
    { value: 'ALL_MEMBERS', label: '회원 전체', description: '등록된 회원 전체' },
    { value: 'ALL_REGISTRANTS', label: '사전등록자 전체', description: '취소 제외 · 결제 여부 무관' },
    { value: 'ALL_SUBMITTERS', label: '초록 접수자 전체', description: '접수 회원 기준 · 임시저장 제외' },
    { value: 'ALL_ACCEPTED', label: '초록 채택자 전체', description: '채택 초록을 보유한 접수 회원' }
] as const;
export type RecipientGroup = typeof RECIPIENT_GROUPS[number]['value'];
